### PR TITLE
check price exists before register an asset in assets module

### PR DESCRIPTION
**Pull request by leonz789**
## Description

<p> When register a new asset in assets module, check if there's already valid price exists for that corresponding asset. The register will fail if the check fail </p>
----

Closes #XXX

## Summary by CodeRabbit

- **New Features**
  - Enhanced asset registration process with price validation and duplicate ID checks.
  - Introduction of the `OracleKeeper` interface for retrieving asset prices.
  - New method in `KeeperOracle` for retrieving specific asset prices.

- **Bug Fixes**
  - Added a new error case for duplicate asset ID registration to improve error handling.

- **Documentation**
  - Updated interfaces and methods for improved clarity in asset and price management.

### CODE REVIEW

File: x/assets/keeper/msg_server.go [hunk 1]
```diff
@@ -54,6 +54,18 @@ func (k Keeper) RegisterClientChain(ctx context.Context, req *assetstype.Registe
 
 func (k Keeper) RegisterAsset(ctx context.Context, req *assetstype.RegisterAssetReq) (*assetstype.RegisterAssetResponse, error) {
 	c := sdk.UnwrapSDKContext(ctx)
+	_, assetID := assetstype.GetStakeIDAndAssetIDFromStr(req.Info.LayerZeroChainID, "", req.Info.Address)
+
+	// once an asset is registered, operator will start trying to update ralted power based on this asset's price, so we have to make sure this asset already has price udpated by oracle-module
```
Comment (TimmyExogenous) [inline x/assets/keeper/msg_server.go]: misspelling: `ralted`

File: x/assets/keeper/msg_server.go [hunk 2]
```diff
@@ -54,6 +54,18 @@ func (k Keeper) RegisterClientChain(ctx context.Context, req *assetstype.Registe
...
+	// TODO: there's no gurantee that the corresponding tokenfeeder is running, the latest price is possible to be some history price. But since currently there's no mechanism to remove an asset from assets module, so we just asume corresponding tokenfeeder will never set endblock for now.
+	if _, err := k.oKeeper.GetSpecifiedAssetsPrice(c, assetID); err != nil {
+		return nil, err
+	}
+
+	if asset, _ := k.GetStakingAssetInfo(c, assetID); asset != nil {
+		c.Logger().Error("register asset with duplicated assetID")
```
Comment (TimmyExogenous) [inline x/assets/keeper/msg_server.go]: It seems like this will influence the correction of mistakes in the registration of asset information.

Comment (leonz789) [reply x/assets/keeper/msg_server.go]: updated

File: x/assets/keeper/msg_server.go (hunk 1 above)
Comment (leonz789) [reply x/assets/keeper/msg_server.go]: updated

**Review by TimmyExogenous — APPROVED** [review]

**Comment by cloud8little** [comment]
Test passed. Register Asset would fail if the token is not registered in oracle module yet, after registering the asset id in oracle module, register asset succeed, and feeding price also works well.

**Review by cloud8little — APPROVED** [review]

**Comment by cloud8little** [comment]
Test Failed. 
Binary:  PR #143 (838d3aef16108550c00b65958bee36628153de0e), PR #125 (https://github.com/ExocoreNetwork/exocore/pull/125/commits/0e4ef8fdfc919d8f5b5ebcb3bcb09e32f0f389ad) + develop (https://github.com/ExocoreNetwork/exocore/commit/18d51364fad619c94b71b2f2273b728342dce0d3) 

```
9:43AM ERR fail to get price from oracle, since current assetID is not bonded with oracle token details:="assetID does not exist in oracle 0x1e867667ef16111047c2c7f6adf4612bdf80064d_0x9cdf: get price failed for asset not found" module=server
9:43AM ERR CONSENSUS FAILURE!!! err="assetID does not exist in oracle 0x1e867667ef16111047c2c7f6adf4612bdf80064d_0x9cdf: get price failed for asset not found" module=consensus server=node stack="goroutine 235 [running]:\nruntime/debug.Stack()\n\truntime/debug/stack.go:24 +0x5e\ngithub.com/cometbft/cometbft/consensus.(*State).receiveRoutine.func2()\n\tgithub.com/cometbft/cometbft@v0.37.4/consensus/state.go:736 +0x46\npanic({0x2f4bdc0?, 0xc0057d4200?})\n\truntime/panic.go:914 +0x21f\ngithub.com/ExocoreNetwork/exocore/x/operator/keeper.(*Keeper).EndBlock(_, {{0x41b7e60, 0x608f580}, {0x41d3610, 0xc005b279c0}, {{0xb, 0x0}, {0xc00433f368, 0x14}, 0x194d, ...}, ...}, ...)\n\tgithub.com/ExocoreNetwork/exocore/x/operator/keeper/abci.go:161 +0x10c\ngithub.com/ExocoreNetwork/exocore/x/operator.AppModule.EndBlock(...)\n\tgithub.com/ExocoreNetwork/exocore/x/operator/module.go:97\ngithub.com/cosmos/cosmos-sdk/types/module.(*Manager).EndBlock(_, {{0x41b7e60, 0x608f580}, {0x41d3610, 0xc005b279c0}, {{0xb, 0x0}, {0xc00433f368, 0x14}, 0x194d, ...}, ...}, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.8/types/module/module.go:583 +0x1c9\ngithub.com/ExocoreNetwork/exocore/app.(*ExocoreApp).EndBlocker(...)\n\tgithub.com/ExocoreNetwork/exocore/app/app.go:1145\ngithub.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).EndBlock(0xc000a981e0, {0x40?})\n\tgithub.com/cosmos/cosmos-sdk@v0.47.8/baseapp/abci.go:219 +0x165\ngithub.com/cometbft/cometbft/abci/client.(*localClient).EndBlockSync(0xc0028bbc80, {0xc005581200?})\n\tgithub.com/cometbft/cometbft@v0.37.4/abci/client/local_client.go:291 +0xd3\ngithub.com/cometbft/cometbft/proxy.(*appConnConsensus).EndBlockSync(0xc001bff050, {0xc0039f5600?})\n\tgithub.com/cometbft/cometbft@v0.37.4/proxy/app_conn.go:111 +0x14c\ngithub.com/cometbft/cometbft/state.execBlockOnProxyApp({0x41b7e98?, 0xc001c97800}, {0x41cf5e0, 0xc001bff050}, 0xc0046ba3c0, {0x41d4348, 0xc000e85080}, 0x194c?)\n\tgithub.com/cometbft/cometbft@v0.37.4/state/execution.go:383 +0x6fc\ngithub.com/cometbft/cometbft/state.(*BlockExecutor).ApplyBlock(_, {{{0xb, 0x0}, {0xc0028cff40, 0x6}}, {0xc0028f4888, 0x14}, 0x1, 0x194c, {{0xc005606220, ...}, ...}, ...}, ...)\n\tgithub.com/cometbft/cometbft@v0.37.4/state/execution.go:197 +0x165\ngithub.com/cometbft/cometbft/consensus.(*State).finalizeCommit(0xc000ee4e00, 0x194d)\n\tgithub.com/cometbft/cometbft@v0.37.4/consensus/state.go:1710 +0xa65\ngithub.com/cometbft/cometbft/consensus.(*State).tryFinalizeCommit(0xc000ee4e00, 0x194d)\n\tgithub.com/cometbft/cometbft@v0.37.4/consensus/state.go:1619 +0x2f6\ngithub.com/cometbft/cometbft/consensus.(*State).enterCommit.func1()\n\tgithub.com/cometbft/cometbft@v0.37.4/consensus/state.go:1554 +0x9c\ngithub.com/cometbft/cometbft/consensus.(*State).enterCommit(0xc000ee4e00, 0x194d, 0x0)\n\tgithub.com/cometbft/cometbft@v0.37.4/consensus/state.go:1592 +0xc5b\ngithub.com/cometbft/cometbft/consensus.(*State).addVote(0xc000ee4e00, 0xc006224a00, {0x0, 0x0})\n\tgithub.com/cometbft/cometbft@v0.37.4/consensus/state.go:2223 +0x18ed\ngithub.com/cometbft/cometbft/consensus.(*State).tryAddVote(0xc000ee4e00, 0xc006224a00, {0x0?, 0x4fd1f0?})\n\tgithub.com/cometbft/cometbft@v0.37.4/consensus/state.go:2012 +0x26\ngithub.com/cometbft/cometbft/consensus.(*State).handleMsg(0xc000ee4e00, {{0x4186020, 0xc00285c580}, {0x0, 0x0}})\n\tgithub.com/cometbft/cometbft@v0.37.4/consensus/state.go:865 +0x3ce\ngithub.com/cometbft/cometbft/consensus.(*State).receiveRoutine(0xc000ee4e00, 0x0)\n\tgithub.com/cometbft/cometbft@v0.37.4/consensus/state.go:792 +0x4bc\ncreated by github.com/cometbft/cometbft/consensus.(*State).OnStart in goroutine 174\n\tgithub.com/cometbft/cometbft@v0.37.4/consensus/state.go:383 +0x10c\n"
```

**Review by TimmyExogenous — APPROVED** [review]

**Comment by cloud8little** [comment]
after applying the latest commit: https://github.com/ExocoreNetwork/exocore/pull/143/commits/cf52fc8374ef2d822e0f7dd942cfef8e787a5f33 , network recovered. ONLY showing error message, not stopping consensus.

```
2:19PM INF ABCI Replay Blocks appHeight=6476 module=consensus server=node stateHeight=6476 storeHeight=6477
2:19PM INF Replay last block using real app module=consensus server=node
2:19PM INF recacheAggregatorContext from=6475 height=6477 module=x/oracle to=6477
2:19PM INF prepare for next oracle round of each tokenFeeder height=6477 module=x/oracle
2:19PM ERR internal error when calling assets precompile err="assetID does not exist in oracle 0x1e867667ef16111047c2c7f6adf4612bdf80064d_0x9cdf: get price failed for asset not found" method=registerToken module="assets precompile"
2:19PM INF executed block height=6477 module=consensus num_invalid_txs=0 num_valid_txs=1 server=node
2:19PM INF commit synced commit=436F6D6D697449447B5B313836203133352031333220313320313136203231203135352032343520323320313734203133342031353520313635203137322032323620323431203137203535203533203734203230312032313120313033203134322037372037382034342034203234302031333220313038203131325D3A313934447D module=server
2:19PM INF committed state app_hash=BA87840D74159BF517AE869BA5ACE2F11137354AC9D3678E4D4E2C04F0846C70 height=6477 module=consensus num_txs=1 server=node
```

File: x/assets/keeper/keeper.go [hunk 3]
```diff
@@ -10,15 +10,18 @@ import (
 type Keeper struct {
 	storeKey storetypes.StoreKey
 	cdc      codec.BinaryCodec
+	assetstype.OracleKeeper
 }
 
 func NewKeeper(
 	storeKey storetypes.StoreKey,
 	cdc codec.BinaryCodec,
+	oKeeper assetstype.OracleKeeper,
```
Comment (bwhour) [inline x/assets/keeper/keeper.go]: let's use oraKeeper / oracleKeeper instead of oKeeper.

**Review by bwhour — APPROVED** [review]

File: x/assets/keeper/keeper.go (hunk 3 above)
Comment (leonz789) [reply x/assets/keeper/keeper.go]: updated

**Review by MaxMustermann2 — APPROVED** [review]

**Pull request merged**